File: src/main/java/tmi/recipe/Recipe.java
package tmi.recipe;

import arc.func.Cons;
import arc.math.Mathf;
import arc.scene.ui.layout.Table;
import arc.struct.ObjectFloatMap;
import arc.struct.OrderedMap;
import arc.util.Nullable;
import tmi.recipe.types.RecipeItem;

import java.util.Objects;

/**配方信息的存储类，该类的每一个实例都表示为一个单独的配方，用于显示配方或者计算生产数据*/
public class Recipe {
  private static final EffFunc ONE_BASE = getDefaultEff(1);
  private static final EffFunc ZERO_BASE = getDefaultEff(0);

  /**该配方的类型，请参阅{@link RecipeType}*/
  public final RecipeType recipeType;
  /**配方的标准耗时，具体来说即该配方在100%的工作效率下执行一次生产的耗时，任意小于0的数字都被认为生产过程是连续的*/
  //meta
  public float time = -1;

  /**配方的产出物列表
   * @see RecipeItemStack*/
  public final OrderedMap<RecipeItem<?>, RecipeItemStack> productions = new OrderedMap<>();
  /**配方的输入材料列表
   * @see RecipeItemStack*/
  public final OrderedMap<RecipeItem<?>, RecipeItemStack> materials = new OrderedMap<>();

  /**配方的效率计算函数，用于给定一个输入环境参数和配方数据，计算出该配方在这个输入环境下的工作效率*/
  public EffFunc efficiency = getOneEff();

  //infos
  /**执行该配方的建筑/方块*/
  public RecipeItem<?> block;
  /**该配方在显示时的附加显示内容构建函数，若不设置则认为不添加任何附加信息*/
  @Nullable public Cons<Table> subInfoBuilder;

  public Recipe(RecipeType recipeType) {
    this.recipeType = recipeType;
  }

  /**用配方当前使用的效率计算器计算该配方在给定的环境参数下的运行效率*/
  public float calculateEfficiency(EnvParameter parameter) {
    return efficiency.calculateEff(this, parameter, calculateMultiple(parameter));
  }

  public float calculateEfficiency(EnvParameter parameter, float multiplier) {
    return efficiency.calculateEff(this, parameter, multiplier);
  }

  public float calculateMultiple(EnvParameter parameter) {
    return efficiency.calculateMultiple(this, parameter);
  }

  //utils

  public RecipeItemStack addMaterial(RecipeItem<?> item, int amount){
    RecipeItemStack res = time > 0?
        new RecipeItemStack(item, amount/time).setIntegerFormat(time):
        new RecipeItemStack(item, amount).setIntegerFormat();

    materials.put(item, res);
    return res;
  }

  public RecipeItemStack addMaterial(RecipeItem<?> item, float amount){
    RecipeItemStack res = time > 0?
        new RecipeItemStack(item, amount/time).setFloatFormat(time):
        new RecipeItemStack(item, amount).setFloatFormat();

    materials.put(item, res);
    return res;
  }

  public RecipeItemStack addMaterialPresec(RecipeItem<?> item, float preSeq){
    RecipeItemStack res = new RecipeItemStack(item, preSeq).setPresecFormat();
    materials.put(item, res);
    return res;
  }

  public RecipeItemStack addMaterialRaw(RecipeItem<?> item, float amount){
    RecipeItemStack res = new RecipeItemStack(item, amount);
    materials.put(item, res);
    return res;
  }

  public RecipeItemStack addProduction(RecipeItem<?> item, int amount){
    RecipeItemStack res = time > 0?
        new RecipeItemStack(item, amount/time).setIntegerFormat(time):
        new RecipeItemStack(item, amount).setIntegerFormat();

    productions.put(item, res);
    return res;
  }

  public RecipeItemStack addProduction(RecipeItem<?> item, float amount){
    RecipeItemStack res = time > 0?
        new RecipeItemStack(item, amount/time).setFloatFormat(time):
        new RecipeItemStack(item, amount).setFloatFormat();

    productions.put(item, res);
    return res;
  }

  public RecipeItemStack addProductionPresec(RecipeItem<?> item, float preSeq){
    RecipeItemStack res = new RecipeItemStack(item, preSeq).setPresecFormat();
    productions.put(item, res);
    return res;
  }

  public RecipeItemStack addProductionRaw(RecipeItem<?> item, float amount){
    RecipeItemStack res = new RecipeItemStack(item, amount);
    productions.put(item, res);
    return res;
  }

  public Recipe setBlock(RecipeItem<?> block){
    this.block = block;
    return this;
  }

  public Recipe setTime(float time){
    this.time = time;
    return this;
  }

  public Recipe setEfficiency(EffFunc func){
    efficiency = func;
    return this;
  }

  public boolean containsProduction(RecipeItem<?> production) {
    return productions.containsKey(production);
  }

  public boolean containsMaterial(RecipeItem<?> material) {
    return materials.containsKey(material);
  }

  /**@see Recipe#getDefaultEff(float) */
  public static EffFunc getOneEff(){
    return ONE_BASE;
  }

  /**@see Recipe#getDefaultEff(float) */
  public static EffFunc getZeroEff(){
    return ZERO_BASE;
  }

  /**生成一个适用于vanilla绝大多数工厂与设备的效率计算器，若{@linkplain RecipeParser 配方解析器}正确的解释了方块，这个函数应当能够正确计算方块的实际工作效率*/
  public static EffFunc getDefaultEff(float baseEff){
    ObjectFloatMap<Object> attrGroups = new ObjectFloatMap<>();

    return new EffFunc() {
      @Override
      public float calculateEff(Recipe recipe, EnvParameter env, float mul) {
        float eff = 1;

        attrGroups.clear();

        for (RecipeItemStack stack : recipe.materials.values()) {
          if (stack.isBooster || stack.isAttribute) continue;

          if (stack.attributeGroup != null){
            float e = attrGroups.get(stack.attributeGroup, 1)*stack.efficiency*Mathf.clamp(env.getInputs(stack.item)/(stack.amount*mul));
            if (stack.maxAttr) {
              attrGroups.put(stack.attributeGroup, Math.max(attrGroups.get(stack.attributeGroup, 0), e));
            }
            else attrGroups.increment(stack.attributeGroup, 0, e);
          }
          else eff *= Math.max(stack.efficiency*Mathf.clamp(env.getInputs(stack.item)/(stack.amount*mul)), stack.optionalCons? 1: 0);
        }

        ObjectFloatMap.Values v = attrGroups.values();
        while (v.hasNext()) {
          eff *= v.next();
        }

        return eff*mul;
      }

      @Override
      public float calculateMultiple(Recipe recipe, EnvParameter env) {
        attrGroups.clear();

        float attr = 0;
        float boost = 1;

        for (RecipeItemStack stack : recipe.materials.values()) {
          if (!stack.isBooster && !stack.isAttribute) continue;

          if (stack.isAttribute){
            float a = stack.efficiency*Mathf.clamp(env.getAttribute(stack.item)/stack.amount);
            if (stack.maxAttr) attr = Math.max(attr, a);
            else attr += a;
          }
          else {
            if (stack.attributeGroup != null){
              float e = attrGroups.get(stack.attributeGroup, 1)*stack.efficiency*Mathf.clamp(env.getInputs(stack.item)/stack.amount);
              if (stack.maxAttr) {
                attrGroups.put(stack.attributeGroup, Math.max(attrGroups.get(stack.attributeGroup, 0), e));
              }
              else attrGroups.increment(stack.attributeGroup, 0, e);
            }
            else boost *= Math.max(stack.efficiency*Mathf.clamp(env.getInputs(stack.item)/stack.amount), stack.optionalCons? 1: 0);
          }
        }

        ObjectFloatMap.Values v = attrGroups.values();
        while (v.hasNext()) {
          boost *= v.next();
        }
        return boost*(baseEff + attr);
      }
    };
  }

  @Override
  public boolean equals(Object object) {
    if (this == object) return true;
    if (!(object instanceof Recipe r)) return false;
    if (r.recipeType != recipeType || r.block != block) return false;

    if (r.materials.size != materials.size || r.productions.size != productions.size) return false;

    return r.materials.equals(materials) && r.productions.equals(productions);
  }

  @Override
  public int hashCode() {
    return Objects.hash(recipeType, productions.orderedKeys(), materials.orderedKeys(), block);
  }

  public interface EffFunc{
    float calculateEff(Recipe recipe, EnvParameter env, float mul);
    float calculateMultiple(Recipe recipe, EnvParameter env);
  }
}

File: src/main/kotlin/tmi/recipe/Recipe.kt
package tmi.recipe

import arc.func.Cons
import arc.scene.ui.layout.Table
import arc.struct.ObjectFloatMap
import arc.struct.OrderedMap
import tmi.util.invoke
import tmi.recipe.types.CalculateMethod
import tmi.recipe.types.RecipeItem
import tmi.recipe.types.RecipeItemType
import tmi.util.set
import tmi.util.mto
import java.util.*
import kotlin.math.max
import kotlin.math.min

/**配方信息的存储类，该类的每一个实例都表示为一个单独的配方，用于显示配方或者计算生产数据
 *
 * @param recipeType [Recipe.recipeType]
 * @param ownerBlock [Recipe.ownerBlock]
 * @param craftTime [Recipe.craftTime]*/
open class Recipe @JvmOverloads constructor(
  /**该配方的类型，请参阅[RecipeType] */
  val recipeType: RecipeType,
  /**该配方从属的方块（可以是执行者/建造目标），具体含义取决于配方的类型*/
  val ownerBlock: RecipeItem<*>,
  /**配方的标准耗时，具体来说即该配方在100%的工作效率下执行一次生产的耗时，任意小于0的数字都被认为生产过程是连续的*/ //meta
  val craftTime: Float = -1f,
) {
  private var completed = false
  private var hash = -1

  private val productionMap = OrderedMap<RecipeItem<*>, RecipeItemStack<*>>()
  private val materialMap = OrderedMap<RecipeItem<*>, RecipeItemStack<*>>()

  val productions get() = productionMap.values().toList()
  val materials get() = materialMap.values().toList()

  val materialGroups: List<List<RecipeItemStack<*>>> get() = run {
    var n = 0
    materials.groupBy { it.group?:n++ }.map { it.value.toList() }
  }

  //infos
  /**[tmi.recipe.types.RecipeItemType.NORMAL]计算区的计算方法
   * @see tmi.recipe.types.CalculateMethod*/
  var normalMethod = CalculateMethod.MIN
    private set
  /**[tmi.recipe.types.RecipeItemType.POWER]计算区的计算方法
   * @see CalculateMethod*/
  var powerMethod = CalculateMethod.MULTIPLE
    private set
  /**[tmi.recipe.types.RecipeItemType.ATTRIBUTE]计算区的计算方法
   * @see CalculateMethod*/
  var attributeMethod = CalculateMethod.MIN
    private set
  /**[tmi.recipe.types.RecipeItemType.BOOSTER]计算区的计算方法
   * @see CalculateMethod*/
  var boosterMethod = CalculateMethod.MIN
    private set
  /**[tmi.recipe.types.RecipeItemType.ISOLATED]计算区的计算方法
   * @see CalculateMethod*/
  var isolatedMethod = CalculateMethod.MIN
    private set

  var baseEfficiency = 1f
    private set

  var subInfoBuilder: Cons<Table>? = null
    private set

  fun complete(){
    if (completed) return
    productionMap.orderedKeys().sort()
    materialMap.orderedKeys().sort()

    hash = Objects.hash(
      recipeType.id,
      productionMap.keys().toList(),
      materialMap.keys().toList(),
      ownerBlock
    )

    completed = true
  }

  /**用配方当前使用的效率计算器计算该配方在给定的环境参数下的运行效率 */
  @JvmOverloads
  open fun calculateEfficiency(parameter: EnvParameter, multiplier: Float = calculateMultiple(parameter)): Float {
    if (multiplier <= 0f) return 0f

    val matN = materials.filter { it.itemType == RecipeItemType.NORMAL }
    val matI = materials.filter { it.itemType == RecipeItemType.ISOLATED }
    val normal = if (matN.any()) calculateZone(
      matN,
      normalMethod,
      parameter,
      multiplier
    ) else 1f
    val booster = calculateZone(
      materials.filter { it.itemType == RecipeItemType.BOOSTER },
      boosterMethod,
      parameter,
      multiplier
    )
    val isolated = if (matI.any()) calculateZone(
      matI,
      isolatedMethod,
      parameter,
      1f
    ) else 1f

    return normal*max(booster, 1f)*multiplier*isolated
  }

  open fun calculateMultiple(parameter: EnvParameter, multiplier: Float = 1f): Float {
    val powerM = materials.filter { it.itemType == RecipeItemType.POWER }
    val attr = calculateZone(
      materials.filter { it.itemType == RecipeItemType.ATTRIBUTE },
      attributeMethod,
      parameter,
      multiplier
    )
    val power = if (powerM.any()) calculateZone(
      powerM,
      powerMethod,
      parameter,
      multiplier
    ) else 1f

    return (baseEfficiency + attr)*power
  }

  internal fun calculateZone(
    filteredStacks: List<RecipeItemStack<*>>,
    method: CalculateMethod,
    parameter: EnvParameter,
    multiplier: Float = 1f
  ): Float {
    val groupEff = ObjectFloatMap<RecipeItemGroup>()

    val effs = filteredStacks.map { it mto it.efficiencyBy(parameter, multiplier) }
    method.cleanOptional(
      effs, effs.filter { it.first.isOptional },
      { it.second }, { second = it }
    )

    effs.forEach {
      if (it.second.isInfinite()) return@forEach
      it.first.group?.also { group ->
        groupEff[group] = max(groupEff.get(group, it.second), it.second)
      }
    }

    val arr = groupEff.values().toArray().also { seq ->
      seq.addAll(*effs.filter { it.first.group == null }.map { it.second }.toFloatArray())
    }

    return method.calculate(arr.toArray())
  }

  //properties
  fun setNormalMethod(attr: CalculateMethod) = also { normalMethod = attr }
  fun setPowerMethod(attr: CalculateMethod) = also { powerMethod = attr }
  fun setAttributeMethod(attr: CalculateMethod) = also { attributeMethod = attr }
  fun setBoosterMethod(attr: CalculateMethod) = also { boosterMethod = attr }
  fun setIsolatedMethod(attr: CalculateMethod) = also { isolatedMethod = attr }
  fun setBaseEff(baseEfficiency: Float) = also { this.baseEfficiency = baseEfficiency }
  fun setSubInfo(builder: Cons<Table>) = also { subInfoBuilder = builder }
  fun prependSubInfo(prependBuilder: Cons<Table>) = also {
    val last = subInfoBuilder
    subInfoBuilder = Cons{
      prependBuilder(it)
      last?.get(it)
    }
  }
  fun appendSubInfo(appendBuilder: Cons<Table>) = also {
    val last = subInfoBuilder
    subInfoBuilder = Cons{
      last?.get(it)
      appendBuilder(it)
    }
  }

  //handles
  fun addMaterial(item: RecipeItem<*>, amount: Number) =
    RecipeItemStack(item, amount.toFloat()).also { materialMap[item] = it }
  fun addMaterial(stack: RecipeItemStack<*>){ materialMap[stack.item] = stack }
  fun addProduction(item: RecipeItem<*>, amount: Number) =
    RecipeItemStack(item, amount.toFloat()).also { productionMap[item] = it }
  fun addProduction(stack: RecipeItemStack<*>){ productionMap[stack.item] = stack }

  fun getMaterial(item: RecipeItem<*>): RecipeItemStack<*>? = materialMap[item]
  fun getProduction(item: RecipeItem<*>): RecipeItemStack<*>? = productionMap[item]

  //utils
  fun addMaterialInteger(item: RecipeItem<*>, amount: Int) =
    (if (craftTime > 0) RecipeItemStack(item, amount/craftTime).integerFormat(craftTime)
    else RecipeItemStack(
      item,
      amount.toFloat()
    ).integerFormat()).also {
      it.setAltFormat(AmountFormatter.timedAmountFormatter())
      materialMap[item] = it
    }

  fun addMaterialFloat(item: RecipeItem<*>, amount: Float) =
    (if (craftTime > 0) RecipeItemStack(item, amount/craftTime).floatFormat(craftTime)
    else RecipeItemStack(
      item,
      amount
    ).floatFormat()).also {
      it.setAltFormat(AmountFormatter.timedAmountFormatter())
      materialMap[item] = it
    }

  fun addMaterialPersec(item: RecipeItem<*>, persec: Float) =
    RecipeItemStack(item, persec).unitTimedFormat().also {
      if (craftTime > 0) it.setAltFormat(AmountFormatter.timedAmountFormatter())
      materialMap[item] = it
    }

  fun addProductionInteger(item: RecipeItem<*>, amount: Int) =
    (if (craftTime > 0) RecipeItemStack(item, amount/craftTime).integerFormat(craftTime)
    else RecipeItemStack(
      item,
      amount.toFloat()
    ).integerFormat()).also {
      it.setAltFormat(AmountFormatter.timedAmountFormatter())
      productionMap[item] = it
    }

  fun addProductionFloat(item: RecipeItem<*>, amount: Float) =
    (if (craftTime > 0) RecipeItemStack(item, amount/craftTime).floatFormat(craftTime)
    else RecipeItemStack(
      item,
      amount
    ).floatFormat()).also {
      it.setAltFormat(AmountFormatter.timedAmountFormatter())
      productionMap[item] = it
    }

  fun addProductionPersec(item: RecipeItem<*>, perSec: Float) =
    RecipeItemStack(item, perSec).unitTimedFormat().also {
      if (craftTime > 0) it.setAltFormat(AmountFormatter.timedAmountFormatter())
      productionMap[item] = it
    }

  fun containsProduction(production: RecipeItem<*>) = productionMap.containsKey(production)
  fun containsMaterial(material: RecipeItem<*>?) = materialMap.containsKey(material)

  fun RecipeItemStack<*>.efficiencyBy(env: EnvParameter, multiplier: Float): Float =
    if (itemType == RecipeItemType.ATTRIBUTE) efficiency*min(env.getAttribute(item)/(amount*multiplier), 1f)
    else efficiency*min(env.getInputs(item)/(amount*multiplier), 1f)

  override fun equals(other: Any?): Boolean {
    if (this === other) return true
    if (other !is Recipe) return false
    if (other.recipeType !== recipeType || other.ownerBlock !== ownerBlock) return false

    if (other.materialMap.size != materialMap.size || other.productionMap.size != productionMap.size) return false

    return other.materialMap == materialMap && other.productionMap == productionMap
  }

  override fun hashCode(): Int {
    if (!completed) throw IllegalStateException("Recipe is not completed")
    return hash
  }

  override fun toString(): String {
    return "recipe(type: $recipeType block: $ownerBlock, time: $craftTime){materials: ${materialMap.orderedKeys()}, productions: ${productionMap.orderedKeys()}}"
  }

  /**配方的效率计算函数，用于给定一个输入环境参数和配方数据，计算出该配方在这个输入环境下的工作效率 */
  @Deprecated("Use standard efficiency calculate method.")
  var efficiencyFunc = object: EffFunc{
    override fun calculateEff(recipe: Recipe, env: EnvParameter, mul: Float) = 0f
    override fun calculateMultiple(recipe: Recipe, env: EnvParameter) = 0f
  }
  @Deprecated("Use standard efficiency calculate method.")
  fun setEff(func: EffFunc) = this.also { efficiencyFunc = func }
  @Deprecated("Use standard efficiency calculate method.")
  interface EffFunc {
    fun calculateEff(recipe: Recipe, env: EnvParameter, mul: Float): Float
    fun calculateMultiple(recipe: Recipe, env: EnvParameter): Float
  }
}
